pkg/services/pluginsintegration/angularinspector/angularinspector.go

```diff
@@ -0,0 +1,25 @@
+package angularinspector
+
+import (
+	"github.com/grafana/grafana/pkg/plugins/config"
+	"github.com/grafana/grafana/pkg/plugins/manager/loader/angular/angularinspector"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+)
+
+type Service struct {
+	angularinspector.Inspector
+}
+
+func ProvideService(cfg *config.Cfg) (*Service, error) {
+	var underlying angularinspector.Inspector
+	var err error
+	if cfg.Features != nil && cfg.Features.IsEnabled(featuremgmt.FlagPluginsDynamicAngularDetectionPatterns) {
+		underlying, err = angularinspector.NewDynamicInspector(cfg)
```

Comment:
you are instantiating the service here in `pluginsintegration` but the code is in `pkg/plugins/manager` (and it's only imported here). My suggestion is to leave only the interface in  `pkg/plugins` and move the implementation code here.

Reply:
I am a bit confused by this and what should live in `pkg/plugins` vs `pluginsintegration`.

The angular inspector itself [doesn't require anything outside of `pkg/plugins`](https://github.com/grafana/grafana/blob/18a6fa861831337bf0571ccb920b42df0fb9d6e3/pkg/plugins/manager/loader/angular/angularinspector/angularinspector.go#L10), however we need `pkg/services/featuremgmt` to switch between the detectors provider depending on feature flag (which is defined in `pkg/services/featuremgmt`), so I have moved the code that switches the implementation in here, and kept the generic inspector implementation in `pkg/plugins` because all its dependencies are in `pkg/plugins`.

I was under the impression that `pluginsintegration` was only for code that depends on other packages outside of `pkg/plugins` (like if it requires kvstore, featuremgmt, pkg/infra, or other Grafana-specific services).

If so, I would prefer leaving the default inspector inside `pkg/plugins`, as the angular detector is required by the plugins loader, and its default implementation also lives in `pkg/plugins`.

@marefr wdyt?

---

I have also noticed that we have [`Features plugins.FeatureToggles` defined in the plugin-specific `Cfg` struct](https://github.com/grafana/grafana/blob/main/pkg/plugins/config/config.go#L41) (it looks like it's unused), so technically it could be used to test for feature flags:

```go
cfg.Features.IsEnabled(featuremgmt.FlagPluginsDynamicAngularDetectionPatterns)
//                     ^~~~~~~~~~~
```

The problem is that we may need to re-define the feature flag if we need to use it `pkg/plugins`, because `featuremgmt` (where the generated feature flags consts are defined) is outside of `pkg/plugins`, and we try to avoid referencing other Grafana packages in `pkg/plugins`.

Reply:
> I am a bit confused by this and what should live in `pkg/plugins` vs `pluginsintegration`.

yeah, I also find it a bit confusing :disappointed: 

> I was under the impression that `pluginsintegration` was only for code that depends on other packages outside of `pkg/plugins` (like if it requires kvstore, featuremgmt, pkg/infra, or other Grafana-specific services).

What I understood is that services `pluginsintegrations` are suitable to be implemented in a different way for other Grafana environments. In theory, that means that we could redefine the inspector to do something else rather than using gcom to retrieve the patterns (or do nothing).

Again, as I understood, in `pkg/plugins` we can define the default behavior. In this case, that would be returning the hardcoded list. Then, in the `pluginsintegrations` we define the service that "optionally" download new patterns dynamically.

As far as Grafana core is concerned, _something_ should implement the inspector interface but the specifics don't need to live in core. Is up to wire to resolve that dependency.

Reply:
Makes sense, I have moved all the gcom-related features (inspector and also patterns provider) to pluginsintegration :smiley: